### ISSUE TITLE
Adding support for Unifi Switches

### ISSUE THREAD

**Issue by tslytsly** [ytti/oxidized#1003]
Hi,

I'm trying to add support for Ubiquiti Unifi Switches.

Essentially they run EdgeSwitch, but the access to the CLI is not the same, Unifi is designed around a Web Controller and CLI access is for debug only.

You need to SSH to the unit and then telnet to localhost, this drops you into the EdgeSwitch CLI.

I have taken the existing EdgeSwitch model config and added the telnet command. (see below)

This works if I use `oxs` but when I add devices to my main oxidized config I get lots of this in the logs:
`WARN -- : <ip_address> raised Timeout::Error with msg "execution expired"`

I think this is because Oxidized is not catching the end of the `show running` command and is waiting forever.
But I'm not sure what to do to find out why it's not seeing the end of the command.

Here is the model file I am using:

``` ruby
class UnifiSwitch < Oxidized::Model

# Ubiquiti UnifiSwitch #

  comment '!'

  prompt /.*[#>]/

  cmd 'show running-config' do |cfg|
    cfg.each_line.to_a[2..-2].reject { |line| line.match /System Up Time.*/ or line.match /Current SNTP Synchronized Time.*/ }.join
  end

  cfg :telnet do
    username /User(name)?:\s?/
    password /^Password:\s?/
  end

  cfg :telnet, :ssh do
    post_login do
      cmd 'telnet localhost'
      if vars :enable
        send "enable\n"
        cmd vars(:enable)
      else
        cmd 'enable'
      end
      cmd 'terminal length 0'
    end
    pre_logout do
      cmd 'quit'
    end

  end
end
```

**Comment by seekerOK** [comment]
debug ?

**Comment by tslytsly** [comment]
Sorry, debug (real hostname and ip removed):

```
D, [2017-08-30T16:32:53.865626 #45785] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for <HostName> at 2017-08-30 15:32:53 UTC
D, [2017-08-30T16:32:53.866157 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to <HostName>
D, [2017-08-30T16:32:53.865561 #45785] DEBUG -- : lib/oxidized/worker.rb: Added <HostName> to the job queue
D, [2017-08-30T16:32:53.867849 #45785] DEBUG -- : lib/oxidized/worker.rb: 2 jobs running in parallel
D, [2017-08-30T16:32:54.455261 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:32:54.868293 #45785] DEBUG -- : lib/oxidized/worker.rb: 2 jobs running in parallel
D, [2017-08-30T16:32:55.265023 #45785] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at <HostName>
D, [2017-08-30T16:32:55.265243 #45785] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: nil, block: #<Proc:0x00000002ec5610@/var/lib/gems/2.3.0/gems/oxidized-0.20.0/lib/oxidized/model/unifiswitch.rb:19> at <HostName>
D, [2017-08-30T16:32:55.265309 #45785] DEBUG -- : lib/oxidized/input/ssh.rb telnet localhost @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:32:55.265531 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:32:55.692993 #45785] DEBUG -- : lib/oxidized/input/ssh.rb enable @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:32:55.693310 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:32:55.869918 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:32:56.299714 #45785] DEBUG -- : lib/oxidized/input/ssh.rb terminal length 0 @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:32:56.300069 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:32:56.704126 #45785] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2017-08-30T16:32:56.704441 #45785] DEBUG -- : lib/oxidized/model/model.rb Executing show running-config
D, [2017-08-30T16:32:56.704511 #45785] DEBUG -- : lib/oxidized/input/ssh.rb show running-config @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:32:56.704688 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:32:56.870255 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:32:57.544887 #45785] DEBUG -- : lib/oxidized/input/cli.rb Running pre_logout commands at <HostName>
D, [2017-08-30T16:32:57.544978 #45785] DEBUG -- : lib/oxidized/input/ssh.rb quit @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:32:57.545164 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
W, [2017-08-30T16:33:17.965580 #45785]  WARN -- : <IP_Address> raised Timeout::Error with msg "execution expired"
D, [2017-08-30T16:33:17.965733 #45785] DEBUG -- : lib/oxidized/node.rb: Oxidized::SSH failed for <HostName>
D, [2017-08-30T16:33:17.965969 #45785] DEBUG -- : lib/oxidized/job.rb: Config fetched for <HostName> at 2017-08-30 15:33:17 UTC
W, [2017-08-30T16:33:18.879808 #45785]  WARN -- : <HostName> status no_connection, retry attempt 1
D, [2017-08-30T16:33:18.879907 #45785] DEBUG -- : lib/oxidized/worker.rb: Jobs 0, Want: 2
D, [2017-08-30T16:33:18.880019 #45785] DEBUG -- : lib/oxidized/worker.rb: Added <HostName> to the job queue
D, [2017-08-30T16:33:18.880052 #45785] DEBUG -- : lib/oxidized/worker.rb: Jobs 1, Want: 2
D, [2017-08-30T16:33:18.880523 #45785] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for <HostName> at 2017-08-30 15:33:18 UTC
D, [2017-08-30T16:33:18.881292 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to <HostName>
D, [2017-08-30T16:33:19.478606 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:20.492350 #45785] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at <HostName>
D, [2017-08-30T16:33:20.493510 #45785] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: nil, block: #<Proc:0x00000002447968@/var/lib/gems/2.3.0/gems/oxidized-0.20.0/lib/oxidized/model/unifiswitch.rb:19> at <HostName>
D, [2017-08-30T16:33:20.494530 #45785] DEBUG -- : lib/oxidized/input/ssh.rb telnet localhost @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:33:20.495499 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:20.882874 #45785] DEBUG -- : lib/oxidized/worker.rb: 2 jobs running in parallel
D, [2017-08-30T16:33:20.902367 #45785] DEBUG -- : lib/oxidized/input/ssh.rb enable @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:33:20.903452 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:21.511526 #45785] DEBUG -- : lib/oxidized/input/ssh.rb terminal length 0 @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:33:21.511963 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:21.885426 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:21.915679 #45785] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2017-08-30T16:33:21.915871 #45785] DEBUG -- : lib/oxidized/model/model.rb Executing show running-config
D, [2017-08-30T16:33:21.915936 #45785] DEBUG -- : lib/oxidized/input/ssh.rb show running-config @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:33:21.916164 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:22.753554 #45785] DEBUG -- : lib/oxidized/input/cli.rb Running pre_logout commands at <HostName>
D, [2017-08-30T16:33:22.753752 #45785] DEBUG -- : lib/oxidized/input/ssh.rb quit @ <HostName> with expect: /.*[#>]/
D, [2017-08-30T16:33:22.753991 #45785] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/.*[#>]/] at <HostName>
D, [2017-08-30T16:33:22.885734 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:23.886061 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:24.886470 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:25.886904 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:26.887250 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:27.887823 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:28.888205 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:29.888621 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:30.889228 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:31.889540 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:32.889849 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:33.890603 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:34.890982 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:35.891511 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:36.891845 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:37.892105 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:38.892449 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:39.892812 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:40.893220 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:41.893821 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-08-30T16:33:42.895192 #45785] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
W, [2017-08-30T16:33:43.177217 #45785]  WARN -- : <IP_Address> raised Timeout::Error with msg "execution expired"
```

**Comment by tslytsly** [comment]
here is a manual example of what the script needs to do:

```
US.v3.8.3# telnet localhost

Entering character mode
Escape character is '^]'.

Warning!
The changes may break controller settings and only be effective until reboot.

(UBNT) >enable

(UBNT) #show running-config
...
...
monitor session 1 destination interface 0/16
monitor session 1 source interface 0/2
monitor session 1 mode
exit


(UBNT) #quitConnection closed by foreign host
US.v3.8.3#
exit
```

**Comment by tslytsly** [comment]
bump...

**Comment by laf** [comment]
Bad prompt detection, matching to early maybe?

**Comment by tslytsly** [comment]
is there any way to delay the logout?

**Comment by tslytsly** [comment]
is there any way to match a second prompt?

What I mean is this, the prompt regex is fairly general, because the initial SSH prompt is `US.v3.8.3#` (although the numbers will change with the software version).
When you telnet into the edgeswitch os it changes to `(UBNT) >` followed by `(UBNT) #` on enable.

Ideally I could do with saying that the prompt will be x and then y.

**Comment by laf** [comment]
I'm not aware of a way to do two prompts. You'd probably have to come up with a way to do it in one.

**Comment by MajesticFalcon** [comment]
If nobody is working on this, I can write it up. I have a UNIFI switch.

**Comment by tslytsly** [comment]
@MajesticFalcon that would be great.
I have several different models, so I'm happy to test.

**Comment by danilopopeye** [comment]
> I'm not aware of a way to do two prompts. You'd probably have to come up with a way to do it in one.

yeah, i think this is the only way that will work today.

@MajesticFalcon let us know if you need any help 😉

**Comment by MajesticFalcon** [comment]
@tslytsl I'll ping you once I start working on this. We are about to turn up our first UNIFI switch in the upcoming months.

**Comment by laf** [comment]
@MajesticFalcon Any update on this?

**Comment by tslytsly** [comment]
@MajesticFalcon did you get chance to look at this?

**Comment by wk** [comment]
@MajesticFalcon @tslytsly is this item still being worked on?

**Comment by simonk83** [comment]
Damn, I'd love this as well

**Comment by MajesticFalcon** [comment]
I never got to this because we moved away from using unifi switches entirely. If you have one you can give me access to I can give it a shot

**Comment by WilliamDEdwards** [comment]
For anyone coming across the `execution expired` error on EdgeSwitches: note that when you're using the non-built-in user (e.g. a dedicated user for Oxidized), `enable` must not be set to `true`, but to the user's password. Otherwise, Oxidized never responds to that prompt, causing a timeout.

**Comment by clifcox** [comment]
Greetings,

I was inspired by the OPs start on this so long ago. So I hammered away on it and it seems to be working ok. Though I didn't know ruby before I started, and my workarounds are a bit Kludgy. Here is what I ended up with:
`
/home/oxidized/.config/oxidized# cat model/unifiswitch.rb
class UnifiSwitch < Oxidized::Model

# Ubiquiti UnifiSwitch #

  comment '!'

  prompt /(!?[\w\s.-]+[#>]\s*)/

  cmd 'show info' do |cfg|
    cfg.gsub! /^.*(System Up Time|show info)(.*?[\n\r])/, ''
    cfg.gsub! /^#{prompt}(.*?[\n\r]?)/, ''
    comment cfg
#   '[' + cfg + ']'
#   cfg
  end

  cmd 'show version' do |cfg|
    cfg.gsub! /^#{prompt}?\s*(show\s+version)?(.*?[\n\r]?)/, ''
    (comment cfg) + "\n\n\n"
#   '[' + cfg + ']' + "\n\n\n"
#   cfg + "\n\n\n"
  end

  cmd 'show running-config' do |cfg|
    cfg.each_line.to_a[2..-2].reject { |line| line.match /System Up Time.*/ or line.match /Current SNTP Synchronized Time.*/ }.join
  end

  cfg :ssh do
    post_login do
       cmd 'cli'
       cmd 'terminal length 0'
    end
    pre_logout do
      cmd 'exit'
      cmd 'exit'
      send "exit\r"
    end

  end
end
`
Maybe some of you have suggestions on how to clean it up a bit. :-)

I noticed that with the first two commands, that often the echo of the command, and or the next prompt ends up in the command output, which is why there is some code to filter those out. Why does that happen?

When in debug mode, I notice that many device log files <IP address>-ssh, or <IP address>-telnet are zero length, even for devices that are successfully providing configs. Why does that happen? Wouldn't you always want debug output if you have it turned on?

In the Oxidized web page, all the devices have unknown in the "Last Changed" column, even though I can look at the list of changes and easily find the last one. What's up with that? ;-)

    Thanks,
    Clif

**Comment by clifcox** [comment]
Greetings @ytti,

I'm a bit new on github. Would you consider reopening this issue for continuity's sake? Or would you prefer I open another and reference back to this one? 

    Thanks for a great project,
    Clif

**Comment by robertcheramy** [comment]
@clifcox open a new issue, or if you want to submit a model, open a PR.
